File: zig/flux_vm.zig
const std = @import("std");

const Op = struct {
    pub const HALT: u8 = 0x00;
    pub const NOP: u8 = 0x01;
    pub const INC: u8 = 0x08;
    pub const DEC: u8 = 0x09;
    pub const NOT: u8 = 0x0A;
    pub const NEG: u8 = 0x0B;
    pub const PUSH: u8 = 0x0C;
    pub const POP: u8 = 0x0D;
    pub const STRIPCONF: u8 = 0x17;
    pub const MOVI: u8 = 0x18;
    pub const ADDI: u8 = 0x19;
    pub const SUBI: u8 = 0x1A;
    pub const ADD: u8 = 0x20;
    pub const SUB: u8 = 0x21;
    pub const MUL: u8 = 0x22;
    pub const DIV: u8 = 0x23;
    pub const MOD: u8 = 0x24;
    pub const AND: u8 = 0x25;
    pub const OR: u8 = 0x26;
    pub const XOR: u8 = 0x27;
    pub const MOV: u8 = 0x3A;
    pub const JNZ: u8 = 0x3D;
    pub const JZ: u8 = 0x3C;
    pub const MOVI16: u8 = 0x40;
    pub const JMP: u8 = 0x43;
    pub const LOOP: u8 = 0x46;
};

pub const NUM_REGS = 64;
pub const STACK_SIZE = 4096;

pub const FluxVM = struct {
    gp: [NUM_REGS]i32,
    stack: [STACK_SIZE]i32,
    sp: i32,
    pc: i32,
    halted: bool,
    cycles: i32,
    strip_n: i32,

    pub fn init() FluxVM {
        return FluxVM{
            .gp = [_]i32{0} ** NUM_REGS,
            .stack = [_]i32{0} ** STACK_SIZE,
            .sp = STACK_SIZE,
            .pc = 0,
            .halted = false,
            .cycles = 0,
            .strip_n = 0,
        };
    }

    pub fn execute(self: *FluxVM, bc: []const u8) i32 {
        while (!self.halted and self.pc < @as(i32, @intCast(bc.len))) {
            const result = self.step(bc);
            if (result < 0) break;
            if (result > 0) self.pc += result;
        }
        return self.cycles;
    }

    fn step(self: *FluxVM, bc: []const u8) i32 {
        if (self.pc >= @as(i32, @intCast(bc.len))) return -1;
        const op = bc[@as(usize, @intCast(self.pc))];
        const sz = formatSize(op);
        if (self.pc + sz > @as(i32, @intCast(bc.len))) { self.halted = true; return -1; }
        self.cycles += 1;
        if (self.strip_n > 0) self.strip_n -= 1;

        switch (op) {
            Op.HALT => { self.halted = true; return 1; },
            Op.NOP => return 1,
            Op.INC => { self.gp[bc[@intCast(self.pc+1)]] += 1; return 2; },
            Op.DEC => { self.gp[bc[@intCast(self.pc+1)]] -= 1; return 2; },
            Op.NOT => { const rd = bc[@intCast(self.pc+1)]; self.gp[rd] = ~self.gp[rd]; return 2; },
            Op.NEG => { const rd = bc[@intCast(self.pc+1)]; self.gp[rd] = -self.gp[rd]; return 2; },
            Op.PUSH => { self.sp -= 1; self.stack[@intCast(self.sp)] = self.gp[bc[@intCast(self.pc+1)]]; return 2; },
            Op.POP => { const rd = bc[@intCast(self.pc+1)]; self.gp[rd] = self.stack[@intCast(self.sp)]; self.sp += 1; return 2; },
            Op.STRIPCONF => { self.strip_n = @intCast(bc[@intCast(self.pc+1)]); return 2; },
            Op.MOVI => { const rd = bc[@intCast(self.pc+1)]; const imm: i8 = @bitCast(bc[@intCast(self.pc+2)]); self.gp[rd] = @intCast(imm); return 3; },
            Op.ADDI => { const rd = bc[@intCast(self.pc+1)]; const imm: i8 = @bitCast(bc[@intCast(self.pc+2)]); self.gp[@intCast(rd)] += @intCast(imm); return 3; },
            Op.SUBI => { const rd = bc[@intCast(self.pc+1)]; const imm: i8 = @bitCast(bc[@intCast(self.pc+2)]); self.gp[@intCast(rd)] -= @intCast(imm); return 3; },
            Op.ADD => { self.gp[bc[@intCast(self.pc+1)]] = self.gp[bc[@intCast(self.pc+2)]] + self.gp[bc[@intCast(self.pc+3)]]; return 4; },
            Op.SUB => { self.gp[bc[@intCast(self.pc+1)]] = self.gp[bc[@intCast(self.pc+2)]] - self.gp[bc[@intCast(self.pc+3)]]; return 4; },
            Op.MUL => { self.gp[bc[@intCast(self.pc+1)]] = self.gp[bc[@intCast(self.pc+2)]] * self.gp[bc[@intCast(self.pc+3)]]; return 4; },
            Op.DIV => {
                if (self.gp[bc[@intCast(self.pc+3)]] == 0) { self.halted = true; return 4; }
                self.gp[bc[@intCast(self.pc+1)]] = @divTrunc(self.gp[bc[@intCast(self.pc+2)]], self.gp[bc[@intCast(self.pc+3)]]);
                return 4;
            },
            Op.MOD => { self.gp[bc[@intCast(self.pc+1)]] = @mod(self.gp[bc[@intCast(self.pc+2)]], self.gp[bc[@intCast(self.pc+3)]]); return 4; },
            Op.AND => { self.gp[bc[@intCast(self.pc+1)]] = self.gp[bc[@intCast(self.pc+2)]] & self.gp[bc[@intCast(self.pc+3)]]; return 4; },
            Op.OR  => { self.gp[bc[@intCast(self.pc+1)]] = self.gp[bc[@intCast(self.pc+2)]] | self.gp[bc[@intCast(self.pc+3)]]; return 4; },
            Op.MOV => { self.gp[bc[@intCast(self.pc+1)]] = self.gp[bc[@intCast(self.pc+2)]]; return 4; },
            Op.JNZ => {
                const rd = bc[@intCast(self.pc+1)];
                const off: i32 = @intCast(@as(i8, @bitCast(bc[@intCast(self.pc+2)])));
                if (self.gp[rd] != 0) { self.pc += off; return 0; }
                return 4;
            },
            Op.JZ => {
                const rd = bc[@intCast(self.pc+1)];
                const off: i32 = @intCast(@as(i8, @bitCast(bc[@intCast(self.pc+2)])));
                if (self.gp[rd] == 0) { self.pc += off; return 0; }
                return 4;
            },
            Op.MOVI16 => {
                const rd = bc[@intCast(self.pc+1)];
                const imm: i16 = @bitCast(@as(u16, bc[@intCast(self.pc+2)]) | (@as(u16, bc[@intCast(self.pc+3)]) << 8));
                self.gp[rd] = @intCast(imm);
                return 4;
            },
            Op.LOOP => {
                const rd = bc[@intCast(self.pc+1)];
                const off: u16 = @as(u16, bc[@intCast(self.pc+2)]) | (@as(u16, bc[@intCast(self.pc+3)]) << 8);
                self.gp[rd] -= 1;
                if (self.gp[rd] > 0) { self.pc -= @intCast(off); return 0; }
                return 4;
            },
            else => return sz,
        }
    }

    fn formatSize(op: u8) i32 {
        return if (op <= 0x07) 1
            else if (op <= 0x0F) 2
            else if (op <= 0x17) 2
            else if (op <= 0x1F) 3
            else if (op <= 0x4F) 4
            else 1;
    }
};

test "halt" {
    var vm = FluxVM.init();
    const bc = [_]u8{Op.HALT};
    _ = vm.execute(&bc);
    try std.testing.expect(vm.halted);
    try std.testing.expect(vm.cycles == 1);
}

test "movi" {
    var vm = FluxVM.init();
    const bc = [_]u8{ Op.MOVI, 0, 42, Op.HALT };
    _ = vm.execute(&bc);
    try std.testing.expect(vm.gp[0] == 42);
}

test "add" {
    var vm = FluxVM.init();
    const bc = [_]u8{ Op.MOVI, 0, 10, Op.MOVI, 1, 20, Op.ADD, 2, 0, 1, Op.HALT };
    _ = vm.execute(&bc);
    try std.testing.expect(vm.gp[2] == 30);
}

test "mul" {
    var vm = FluxVM.init();
    const bc = [_]u8{ Op.MOVI, 0, 7, Op.MOVI, 1, 6, Op.MUL, 2, 0, 1, Op.HALT };
    _ = vm.execute(&bc);
    try std.testing.expect(vm.gp[2] == 42);
}

test "inc_dec" {
    var vm = FluxVM.init();
    const bc = [_]u8{ Op.MOVI, 0, 10, Op.INC, 0, Op.INC, 0, Op.DEC, 0, Op.HALT };
    _ = vm.execute(&bc);
    try std.testing.expect(vm.gp[0] == 11);
}

test "push_pop" {
    var vm = FluxVM.init();
    const bc = [_]u8{ Op.MOVI, 0, 42, Op.PUSH, 0, Op.MOVI, 0, 0, Op.POP, 1, Op.HALT };
    _ = vm.execute(&bc);
    try std.testing.expect(vm.gp[1] == 42);
}

test "movi16" {
    var vm = FluxVM.init();
    const bc = [_]u8{ Op.MOVI16, 0, 0xE8, 0x03, Op.HALT };
    _ = vm.execute(&bc);
    try std.testing.expect(vm.gp[0] == 1000);
}

test "fibonacci" {
    var vm = FluxVM.init();
    const bc = [_]u8{
        Op.MOVI, 0, 1,  Op.MOVI, 1, 1,  Op.MOVI, 2, 10,
        Op.ADD, 3, 0, 1,  Op.MOV, 0, 1, 0,  Op.MOV, 1, 3, 0,
        Op.DEC, 2,  Op.JNZ, 2, @bitCast(@as(i8, -14)), 0,
        Op.HALT
    };
    _ = vm.execute(&bc);
    try std.testing.expect(vm.gp[1] == 144);
}

const expect = std.testing.expect;

File: js/flux_vm.js
const OP = {
  HALT:0x00,NOP:0x01,INC:0x08,DEC:0x09,NOT:0x0A,NEG:0x0B,
  PUSH:0x0C,POP:0x0D,STRIPCONF:0x17,
  MOVI:0x18,ADDI:0x19,SUBI:0x1A,
  ADD:0x20,SUB:0x21,MUL:0x22,DIV:0x23,MOD:0x24,
  AND:0x25,OR:0x26,XOR:0x27,
  MIN:0x2A,MAX:0x2B,
  CMP_EQ:0x2C,CMP_LT:0x2D,CMP_GT:0x2E,CMP_NE:0x2F,
  MOV:0x3A,JZ:0x3C,JNZ:0x3D,
  MOVI16:0x40,JMP:0x43,LOOP:0x46,
};

class FluxVM {
  constructor() {
    this.gp = new Int32Array(64);
    this.conf = new Int32Array(64);
    this.stack = new Int32Array(4096);
    this.sp = 4096;
    this.pc = 0;
    this.halted = false;
    this.cycles = 0;
    this.stripN = 0;
  }
  execute(bc) {
    while (!this.halted && this.pc >= 0 && this.pc < bc.length) {
      const r = this.step(bc);
      if (r < 0) break;
      if (r > 0) this.pc += r;
    }
    return this.cycles;
  }
  static fmtSz(op) {
    if (op <= 0x07) return 1;
    if (op <= 0x17) return 2;
    if (op <= 0x1F) return 3;
    if (op <= 0x4F) return 4;
    return 1;
  }
  step(bc) {
    if (this.pc >= bc.length) return -1;
    const op = bc[this.pc];
    const sz = FluxVM.fmtSz(op);
    if (this.pc + sz > bc.length) { this.halted = true; return -1; }
    this.cycles++;
    if (this.stripN > 0) this.stripN--;
    const rd = (i) => bc[this.pc + i];
    switch (op) {
      case OP.HALT: this.halted = true; return 1;
      case OP.NOP: return 1;
      case OP.INC: this.gp[rd(1)]++; return 2;
      case OP.DEC: this.gp[rd(1)]--; return 2;
      case OP.NOT: this.gp[rd(1)] = ~this.gp[rd(1)]; return 2;
      case OP.NEG: this.gp[rd(1)] = -this.gp[rd(1)]; return 2;
      case OP.PUSH: this.stack[--this.sp] = this.gp[rd(1)]; return 2;
      case OP.POP: this.gp[rd(1)] = this.stack[this.sp++]; return 2;
      case OP.STRIPCONF: this.stripN = rd(1); return 2;
      case OP.MOVI: this.gp[rd(1)] = (bc[this.pc+2] << 24) >> 24; return 3;
      case OP.ADDI: this.gp[rd(1)] += (bc[this.pc+2] << 24) >> 24; return 3;
      case OP.SUBI: this.gp[rd(1)] -= (bc[this.pc+2] << 24) >> 24; return 3;
      case OP.ADD: this.gp[rd(1)] = this.gp[rd(2)] + this.gp[rd(3)]; return 4;
      case OP.SUB: this.gp[rd(1)] = this.gp[rd(2)] - this.gp[rd(3)]; return 4;
      case OP.MUL: this.gp[rd(1)] = this.gp[rd(2)] * this.gp[rd(3)]; return 4;
      case OP.DIV:
        if (this.gp[rd(3)] === 0) { this.halted = true; return 4; }
        this.gp[rd(1)] = Math.trunc(this.gp[rd(2)] / this.gp[rd(3)]); return 4;
      case OP.MOD: this.gp[rd(1)] = this.gp[rd(2)] % this.gp[rd(3)]; return 4;
      case OP.AND: this.gp[rd(1)] = this.gp[rd(2)] & this.gp[rd(3)]; return 4;
      case OP.OR:  this.gp[rd(1)] = this.gp[rd(2)] | this.gp[rd(3)]; return 4;
      case OP.XOR: this.gp[rd(1)] = this.gp[rd(2)] ^ this.gp[rd(3)]; return 4;
      case OP.MIN: this.gp[rd(1)] = Math.min(this.gp[rd(2)], this.gp[rd(3)]); return 4;
      case OP.MAX: this.gp[rd(1)] = Math.max(this.gp[rd(2)], this.gp[rd(3)]); return 4;
      case OP.CMP_EQ: this.gp[rd(1)] = (this.gp[rd(2)] === this.gp[rd(3)]) ? 1 : 0; return 4;
      case OP.CMP_LT: this.gp[rd(1)] = (this.gp[rd(2)] < this.gp[rd(3)]) ? 1 : 0; return 4;
      case OP.CMP_GT: this.gp[rd(1)] = (this.gp[rd(2)] > this.gp[rd(3)]) ? 1 : 0; return 4;
      case OP.CMP_NE: this.gp[rd(1)] = (this.gp[rd(2)] !== this.gp[rd(3)]) ? 1 : 0; return 4;
      case OP.MOV: this.gp[rd(1)] = this.gp[rd(2)]; return 4;
      case OP.JZ:  if (this.gp[rd(1)] === 0) { this.pc += ((bc[this.pc+2] << 24) >> 24); return 0; } return 4;
      case OP.JNZ: if (this.gp[rd(1)] !== 0) { this.pc += ((bc[this.pc+2] << 24) >> 24); return 0; } return 4;
      case OP.MOVI16: {
        const imm = bc[this.pc+2] | (bc[this.pc+3] << 8);
        this.gp[rd(1)] = imm > 0x7FFF ? imm - 0x10000 : imm;
        return 4;
      }
      case OP.JMP: {
        let off = bc[this.pc+2] | (bc[this.pc+3] << 8);
        if (off > 0x7FFF) off -= 0x10000;
        this.pc += off; return 0;
      }
      case OP.LOOP: {
        const r = rd(1);
        const off = bc[this.pc+2] | (bc[this.pc+3] << 8);
        this.gp[r]--;
        if (this.gp[r] > 0) { this.pc -= off; return 0; }
        return 4;
      }
      default: return sz;
    }
  }
}

// Tests
const assert = (n, c) => { if(c){console.log(`  ${n.padEnd(50)}PASS`);return 1;} console.log(`  ${n.padEnd(50)}FAIL`);return 0; };
const bc = (...v) => new Uint8Array(v);
let p=0;
console.log('\nFLUX JavaScript VM Tests\n==================\n');
{const v=new FluxVM();v.execute(bc(0x00));p+=assert('halt',v.halted&&v.cycles===1);}
{const v=new FluxVM();v.execute(bc(0x18,0,42,0x00));p+=assert('movi',v.gp[0]===42);}
{const v=new FluxVM();v.execute(bc(0x18,0,0xFB,0x00));p+=assert('movi_neg',v.gp[0]===-5);}
{const v=new FluxVM();v.execute(bc(0x40,0,0xE8,0x03,0x00));p+=assert('movi16',v.gp[0]===1000);}
{const v=new FluxVM();v.execute(bc(0x18,0,10,0x18,1,20,0x20,2,0,1,0x00));p+=assert('add',v.gp[2]===30);}
{const v=new FluxVM();v.execute(bc(0x18,0,7,0x18,1,6,0x22,2,0,1,0x00));p+=assert('mul',v.gp[2]===42);}
{const v=new FluxVM();v.execute(bc(0x18,0,42,0x18,1,7,0x23,2,0,1,0x00));p+=assert('div',v.gp[2]===6);}
{const v=new FluxVM();v.execute(bc(0x18,0,10,0x08,0,0x08,0,0x09,0,0x00));p+=assert('inc_dec',v.gp[0]===11);}
{const v=new FluxVM();v.execute(bc(0x18,0,42,0x0C,0,0x18,0,0,0x0D,1,0x00));p+=assert('push_pop',v.gp[1]===42);}
{const v=new FluxVM();v.execute(bc(0x18,0,5,0x18,1,0,0x08,1,0x09,0,0x3D,0,0xFC,0,0x00));p+=assert('jnz_loop',v.gp[0]===0&&v.gp[1]===5);}
{const v=new FluxVM();v.execute(bc(0x18,0,1,0x18,1,1,0x18,2,10,0x20,3,0,1,0x3A,0,1,0,0x3A,1,3,0,0x09,2,0x3D,2,0xF2,0,0x00));p+=assert('fibonacci',v.gp[1]===144);}
console.log(`\n==================\nResults: ${p} passed, ${11-p} failed\n`);
